### DISCUSSION TITLE
modules vs. driver

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#21218] [Q&A]
Can someone differentiate between modules and drivers? I know that drivers respond to Tasmota messages by calling the corresponding predefined function names. But can I also create individual methods in drivers and include and call them from outside, just like a module?

**Comment by sfromis** [comment]
Drivers and modules are not related concepts.

A driver is when a class instance is registered with `tasmota.add_driver`, meaning that methods with [predefined names](https://tasmota.github.io/docs/Berry/#creating-a-tasmota-driver) will be automatically called for certain events, like `every_250ms`. This class (which it needs to be) can be in a module or not, but a driver as such is nothing like a module. A driver is just a normal class, with the extra detail that Tasmota will call the driver methods when relevant.

A module is to create an encapsulated unit of code with a defined API, and loading can be managed automatically when the code use `import` wherever relevant. Tasmota manages a table of already imported modules. Creating a module includes making selected elements available outside of the module, with the rest been hidden inside. It can work very well for creating modules working in a similar way as the modules Tasmota already provides with Berry. Modules can include managing a state for the feature as such, or it can produce instances the outside code can control.

**Reply by ChrisVonFrosta** [reply]
oh understand. A driver isn't a file like module. It is only the class that has one or more predefine named methods inside and is registered as driver. 
But could has this Class more methods, that not related to the driver and beeing inside a module?

**Reply by sfromis** [reply]
The class is the starting point, and if it needs feature like a web UI button (or whatever), being a driver is pretty useful. Thus the class would typically have a multitude of methods to do it's job, and some of these may be driver methods. All methods share the same instance variables, to act on the main goal of the class. Do not think of a class as **being** a driver, that is merely a strategy for integrating with Tasmota, as part of it's functions.
